File: src/main/java/ca/rafemurray/ModBlocks.java
package ca.rafemurray;

import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents;
import net.minecraft.block.AbstractBlock;
import net.minecraft.block.Block;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.util.Identifier;

import java.util.function.Function;

public class ModBlocks {
    private static Block register(String name, Function<AbstractBlock.Settings, Block> blockFactory, AbstractBlock.Settings settings, boolean registerItem) {
        RegistryKey<Block> blockKey = getBlockKey(name);
        Block block = blockFactory.apply(settings.registryKey(blockKey));

        if (registerItem) {
            RegistryKey<Item> itemKey = getItemKey(name);
            BlockItem item = new BlockItem(block, new Item.Settings().registryKey(itemKey));
            Registry.register(Registries.ITEM, itemKey, item);
        }

        return Registry.register(Registries.BLOCK, blockKey, block);
    }

    private static RegistryKey<Item> getItemKey(String name) {
        return RegistryKey.of(RegistryKeys.ITEM, Identifier.of(ElytraRacing.MOD_ID, name));
    }

    private static RegistryKey<Block> getBlockKey(String name) {
        return RegistryKey.of(RegistryKeys.BLOCK, Identifier.of(ElytraRacing.MOD_ID, name));
    }

    public static final Block BOOSTER = register("booster", BoosterBlock::new, AbstractBlock.Settings.create().noCollision(),true);
    public static void initialize() {
        ItemGroupEvents.modifyEntriesEvent(ModItems.CUSTOM_ITEM_GROUP_KEY).register((itemGroup) -> {
            itemGroup.add(ModBlocks.BOOSTER.asItem());
        });
    }


}

File: src/main/java/ca/rafemurray/block/ModBlocks.kt
package ca.rafemurray.block

import ca.rafemurray.ElytraRacing
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.resources.Identifier
import net.minecraft.resources.ResourceKey
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockBehaviour
import java.util.function.Function

object ModBlocks {
    fun register(
        name: String,
        blockFactory: Function<BlockBehaviour.Properties, Block>,
        settings: BlockBehaviour.Properties,
        shouldRegisterItem: Boolean
    ): Block {
        // Create a registry key for the block
        val blockKey = keyOfBlock(name)
        // Create the block instance
        val block = blockFactory.apply(settings.setId(blockKey))

        // Sometimes, you may not want to register an item for the block.
        // Eg: if it's a technical block like `minecraft:moving_piston` or `minecraft:end_gateway`
        if (shouldRegisterItem) {
            // Items need to be registered with a different type of registry key, but the ID
            // can be the same.
            val itemKey = keyOfItem(name)

            val blockItem = BlockItem(block, Item.Properties().setId(itemKey).useBlockDescriptionPrefix())
            Registry.register(BuiltInRegistries.ITEM, itemKey, blockItem)
        }

        return Registry.register(BuiltInRegistries.BLOCK, blockKey, block)
    }

    private fun keyOfBlock(name: String): ResourceKey<Block> {
        return ResourceKey.create(Registries.BLOCK, Identifier.fromNamespaceAndPath(ElytraRacing.MOD_ID, name))
    }

    private fun keyOfItem(name: String): ResourceKey<Item> {
        return ResourceKey.create(Registries.ITEM, Identifier.fromNamespaceAndPath(ElytraRacing.MOD_ID, name))
    }

    val BOOSTER: Block = register(
        "booster",
        { properties: BlockBehaviour.Properties -> BoosterBlock(properties) },
        BlockBehaviour.Properties.of().noCollision(),
        true
    )

    val START: Block = register(
        "start_block",
        { settings: BlockBehaviour.Properties ->
            StartBlock(settings)
        }, BlockBehaviour.Properties.of(), true
    )

    val FINISH: Block = register(
        "finish_block",
        { settings: BlockBehaviour.Properties -> FinishBlock(settings) }, BlockBehaviour.Properties.of(), true
    )

    val CHECKPOINT: Block = register(
        "checkpoint_block",
        { settings: BlockBehaviour.Properties -> CheckpointBlock(settings) }, BlockBehaviour.Properties.of(), true
    )

    fun initialize() {}
}
